### ISSUE TITLE
Use destructors for automaticaly closing of resources

### ISSUE THREAD

**Issue by b3liever** [nim-lang/cairo#13]
in order to do these are the steps that need to be taken:

1. types and procs interfacing directly with libcairo need to keep their original names
[example type](https://www.cairographics.org/manual/cairo-cairo-surface-t.html#cairo-surface-t) [example proc](https://www.cairographics.org/manual/cairo-Image-Surfaces.html#cairo-image-surface-create)

2. The "higher level" user api of this library will get shorter names like ``Surface``, ``imageSurfaceCreate``.

3. destructors will be attached to these types. it will look like this:

```nim
type
  Surface = object
    data: PSurface

`=destroy`(surface: var Surface)
`=`(dst: var Surface, src: Surface) {.error.}
```

**Comment by Araq** [comment]
I like it but maybe it should be under ``nim-lang/cairo2``?

**Comment by treeform** [comment]
This would not be a big win for me. Usually one creates 1 surface for the entire app. Its not hard to destroy it at the end of your program ... or event just forget and quit and have the OS do it for you.

**Comment by Araq** [comment]
@treeform But are surfaces the only exposed resource? The surface is only an example here.

**Comment by treeform** [comment]
@araq Here are all of the destroys in the library:
```nim
proc destroy*(cr: ptr Context)
proc destroy*(options: ptr FontOptions)
proc destroy*(font_face: ptr FontFace)
proc destroy*(scaled_font: ptr ScaledFont)
proc destroy*(path: ptr Path)
proc destroy*(surface: ptr Surface)
proc destroy*(pattern: ptr Pattern)
```
I never had an issue with any of them, so I feel like it will not impact me in any way. It might be a cool thing to do, even if it's not a useful thing to do.

Will the `=destroy` only work with `--gc:arc`?

**Comment by b3liever** [comment]
I made a minimal [example](https://gist.github.com/b3liever/3e1adf194ad22246f5354a6c1fe891a0)

**Comment by Araq** [comment]
> Will the =destroy only work with --gc:arc?

No, it's always active, however the old seq implementation (and stuff building on top like tables) ignores destructors.

**Comment by b3liever** [comment]
Fork exists at https://github.com/b3liever/cairo2 @treeform try it if you like

**Comment by StefanSalewski** [comment]
Don't forget that we have some sort of high level cairo in gintro for some years now. I have not spent too much effort in it yet, as it does not support introspection and toggle_references, so it is different from the other gtk stuff. But it seems to work with ARC now. I don't accept pull requests, but I may accepts issues or suggestions like the "=" proc to create errors. May add it soon, also for GTK widgets and that.
